zerver/lib/thumbnail.py

```diff
@@ -0,0 +1,84 @@
+# -*- coding: utf-8 -*-
+from __future__ import absolute_import
+
+import os
+import sys
+from six.moves import urllib
+from django.conf import settings
+from typing import Text
+
+ZULIP_PATH = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath('__file__'))))
+sys.path.append(ZULIP_PATH)
+
+from zthumbor.loaders.helpers import (
+    get_sign_hash, THUMBOR_S3_TYPE, THUMBOR_LOCAL_FILE_TYPE,
+    THUMBOR_EXTERNAL_TYPE
+)
+from zerver.lib.upload import get_signed_upload_url
+from zerver.lib.camo import get_camo_url
+
+def is_thumbor_enabled():
+    # type: () -> bool
+    return settings.THUMBOR_HOST != ''
+
+def get_source_type(url):
+    # type: (str) -> str
+    if not url.startswith('user_uploads/'):
+        return THUMBOR_EXTERNAL_TYPE
+
+    local_uploads_dir = settings.LOCAL_UPLOADS_DIR
+    if local_uploads_dir:
+        return THUMBOR_LOCAL_FILE_TYPE
+    return THUMBOR_S3_TYPE
+
+def generate_thumbnail_url(path, size='0x0'):
+    # type: (str, str) -> Text
+    if not is_thumbor_enabled():
+        if path.startswith('https'):
+            return path
+        if path.startswith('http'):
+            return get_camo_url(path)
+        return '/' + path
+
+    # Ignore thumbnailing for static resources.
+    if path.startswith('static/'):
+        return '/' + path
+
+    is_files = ''
+    source_type = get_source_type(path)
+
+    if source_type == THUMBOR_LOCAL_FILE_TYPE:
+        url = path[len('user_uploads/'):]
+        is_files = '/files'
+    if source_type == THUMBOR_S3_TYPE:
+        url = get_signed_upload_url(path)
+    if source_type == THUMBOR_EXTERNAL_TYPE:
+        url = path
+
+    # Hack to get by weird issue with http in path and nginx.
+    # We remove http and https from starting of urls.
+    if url.startswith('https://'):
+        url = url[len('https://'):]
+    if url.startswith('http://'):
+        url = url[len('http://'):]
+    if url.startswith('https:/'):
+        url = url[len('https:/'):]
+    if url.startswith('http:/'):
+        url = url[len('http:/'):]
```

Comment:
Do you understand this hack @adnrs96?

Reply:
consider a situation as follows:
There is a link to an external image say `https://www.google.com/images/srpr/logo4w.png`. When pasted and sent as a message the rendered body of the message will have a block for inline image which will use relative url to refer to this image as: `/thumbnail/https://www.google.com/images/srpr/logo4w.png`. Idea behind the endpoint is any link sent to the endpoint by doing something like `/thumbnail/<link-here>` will get thumbnailed by thumbor.

Thing is when this kind of links hit Nginx, nginx misbehaves.
>What type of link?
>>Links with `https://` or `http://` in there path

So my best guess here is Nginx finds it difficult to parse such links and ends up miss-redirecting such requests. So I placed in this code (or what I like to call hack because it sounds cool!) to handle nginx's misbehaviour.
@timabbott Sorry for the super delayed explanation on this. I thought I already had replied to this question back when you asked.

Reply:
So, first, `https:/` URLs don't really exist, so I'm not sure why you'd need the second half.  But also, this can be done non-hackisly using the `urllib.parse` library -- you just want to get rid of the `netloc` section.